overlord/exportstate/special.go

```diff
@@ -0,0 +1,191 @@
+}
+
+// manifestForCoreSystem returns the manifest of the host as seen on ubuntu-core systems.
+func manifestForCoreSystem() *Manifest {
+	return &Manifest{
+		ExportedForSnapdAsVersion: "host", // Exception from the rule
+		SourceIsHost:              true,
+		// There are no export sets here, snapd snap is going to provide the tools.
+	}
+}
+
+// manifestForClassicSystem returns the manifest of the host as seen on classic systems.
+func manifestForClassicSystem() *Manifest {
+	tools := snapdTools(dirs.DistroLibExecDir)
+	return &Manifest{
+		ExportedForSnapdAsVersion: "host", // Exception from the rule
+		SourceIsHost:              true,
+		Sets:                      map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForSnapdSnap returns the manifest of the snapd snap.
+func manifestForSnapdSnap(info *snap.Info) *Manifest {
+	tools := snapdTools("/usr/lib/snapd")
+	return &Manifest{
+		SnapInstanceName: info.InstanceName(),
+		SnapRevision:     info.Revision,
+		Sets:             map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForCoreSnap returns the manifest of the core snap.
+func manifestForCoreSnap(info *snap.Info) *Manifest {
+	tools := snapdTools("/usr/lib/snapd")
+	return &Manifest{
+		SnapInstanceName:          info.InstanceName(),
+		SnapRevision:              info.Revision,
+		ExportedForSnapdAsVersion: exportedForSnapdAsVersionForCore(info), // Exception from the rule
+
+		// Separate to avoid gofmt annoyance across versions.
+		Sets: map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForRegularSnap returns the manifest for a snap other than core or snapd.
+func manifestForRegularSnap(info *snap.Info) *Manifest {
+	return &Manifest{
+		SnapInstanceName: info.InstanceName(),
+		SnapRevision:     info.Revision,
+		// TODO: eventually get this from the snap.yaml
+	}
+}
+
+func exportedForSnapdAsVersionForCore(info *snap.Info) string {
+	return fmt.Sprintf("core_%s", info.Revision)
+}
+
+func exportedNameVersionForRegularSnap(info *snap.Info) (exportedName string, exportedVersion string) {
+	if typ := info.Type(); typ == snap.TypeSnapd || typ == snap.TypeOS {
+		panic(fmt.Sprintf("internal error, cannot use exportedNameVersionForRegularSnap with core or snapd, got %q of type %q", info.SnapName(), info.Type()))
+	}
+	return info.InstanceName(), info.SnapRevision().String()
+}
+
+func effectiveExportedVersionForSnapdOrCore(st *state.State) (exportedVersion string, err error) {
+	snapdInfo, coreInfo, err := currentSnapdAndCoreInfo(st)
+	if err != nil {
+		return "", err
+	}
+	var activeSnapdExportedVersion string
+	var activeCoreExportedVersion string
+	if snapdInfo != nil && snapdInfo.Broken == "" {
+		activeSnapdExportedVersion = snapdInfo.Revision.String()
+	}
+	if coreInfo != nil && coreInfo.Broken == "" {
+		activeCoreExportedVersion = exportedForSnapdAsVersionForCore(coreInfo)
+	}
+	exportedVersion = selectExportedVersionForSnapdTools(activeSnapdExportedVersion, activeCoreExportedVersion)
```

Comment:
shouldn't this return an error if exportedVersion is "" ?  the caller will have to do something reasonable with it though

Reply:
Not necessarily, on core, before we seed, there is no version that is exported. The result is only used in two places:

1) in exportSnapdTools, which passes it to updateExportedVersion, which handles the empty value as a sign that current symlink should not exist
2) in ExportedNameVersion which just returns it

The only usage of ExportedNameVersion is in the link participant which, again, passes it to updateExportedVersion.

As such I think this is fine.